### PR TITLE
Feature/moving error to framework

### PR DESCRIPTION
**Pull request by reyoung**
Move `paddle/utils/Error.h` to `paddle/framework/error.h`.

Make new framework using `paddle::Error` as error type.

### CODE REVIEW

File: paddle/framework/error.h [hunk 1]
```diff
@@ -122,15 +117,13 @@ class Error {
...
-  /**
-   * @brief check this status by glog.
-   * @note It is a temp method used during cleaning Paddle code. It will be
-   *       removed later.
-   */
-  void check() const { CHECK(this->isOK()) << msg(); }
-
-private:
+ private:
   std::shared_ptr<std::string> msg_;
```
Comment (typhoonzero) [inline paddle/framework/error.h]: Why a `shared_ptr`, can ::std::string object do the job?

Comment (reyoung) [inline paddle/framework/error.h]: No. Because if we using `std:: string`, `Error` will copy by value each time, which is very time-consuming.

Comment (typhoonzero) [inline paddle/framework/error.h]: Well, `Error` object creation will do one memory operation. But after that, we'll pass error object using reference all the time, we can also use `Error` as return types of functions as we have [return value optimization](https://en.wikipedia.org/wiki/Return_value_optimization)

Comment (reyoung) [inline paddle/framework/error.h]: We should not mainly use `foo(Error* error)` to return Error. `__must_check` will let developer force to check the return error, but `foo(Error* error)` does not have the same technic. So it is better to use `Error __must_check foo();`.

In another hand, [RVO](https://en.wikipedia.org/wiki/Return_value_optimization) depends on compiler implementation.

File: paddle/framework/error.h [hunk 2]
```diff
@@ -122,15 +117,23 @@ class Error {
    */
   bool isOK() const { return msg_ == nullptr; }
 
-  /**
-   * @brief check this status by glog.
-   * @note It is a temp method used during cleaning Paddle code. It will be
-   *       removed later.
-   */
-  void check() const { CHECK(this->isOK()) << msg(); }
+  bool operator==(const Error &o) const {
```
Comment (reyoung) [inline paddle/framework/error.h]: @typhoonzero By using `msg_` a shared_ptr, the `operator==` will be fast for predefined Error.

File: paddle/framework/error.h (hunk 1 above)
Comment (typhoonzero) [inline paddle/framework/error.h]: I'm thinking of two type of use case, which is already in this PR:

```c++
// single error return value
Error err = foo();
// multiple return values with Error
Error err;
ReturnValue ret;
foo(&err, &ret);
```

File: paddle/framework/error.h (hunk 2 above)
Comment (typhoonzero) [inline paddle/framework/error.h]: I see.

File: paddle/framework/error.h [hunk 3]
```diff
@@ -122,15 +117,23 @@ class Error {
    */
   bool isOK() const { return msg_ == nullptr; }
```
Comment (helinwang) [inline paddle/framework/error.h]: I am not very familiar with C++ naming convention so I could be wrong: to me it seems `ok()` is better than `isOK()`, "is" makes develop type two more characters, and does not provide much value in my opinion.

**Comment by helinwang** [comment]
Do we need error code to specify what kind of error it is? Two errors with different error message could be the same kind of error (e.g., `Error("open file failed: /tmp/foo")`, `Error("open file failed: /tmp/bar")`).

File: paddle/framework/error.h (hunk 3 above)
Comment (reyoung) [inline paddle/framework/error.h]: My fault, It will be changed soon.

**Comment by reyoung** [comment]
> Do we need error code to specify what kind of error it is? Two errors with different error message could be the same kind of error (e.g., Error("open file failed: /tmp/foo"), Error("open file failed: /tmp/bar")).

In this case, maybe a global function `bool IsNotFound(const Error& err)` is need.

```cpp
bool IsNotFound(const Error& err) {
   auto pattern = r"file not found: .*";
   return pattern.math(err.msg());
}
```

`IsNotFound` is slower than Error Code but it may not be a performance critical path for Paddle.

And I think the most error of Paddle should not be a dynamic string. The most error should be just a predefined const string.

```cpp
const Error FileNotFound = Error("file is not found.");
```

Then `IsNotFound` should just invoke error's `operator ==`, and just a pointer match operation.

File: paddle/framework/error.h (hunk 3 above)
Comment (reyoung) [inline paddle/framework/error.h]: Done.

**Comment by reyoung** [comment]
Closed because #2646

**Pull request closed without merging**